usersync/ejector.go

```diff
@@ -0,0 +1,134 @@
+package usersync
+
+import (
+	"errors"
+	"math"
+	"time"
+)
+
+type Ejector interface {
+	Choose(uids map[string]UIDEntry) (string, error)
+}
+
+type OldestEjector struct {
+	nonPriorityKeys []string
+	PriorityGroups  [][]string
+	EjectPriority   bool
+}
+
+type PriorityBidderEjector struct {
+	PriorityGroups   [][]string
+	SyncerKey        string
+	OldestEjector    OldestEjector
+	IsSyncerPriority bool
+}
+
+// Choose method for oldest ejector will return the oldest uid after determing which set of elements to be choosing from
+func (o *OldestEjector) Choose(uids map[string]UIDEntry) (string, error) {
+	var elements []string
+
+	// Set which elements the ejector will be choosing from
+	if len(o.nonPriorityKeys) > 0 {
+		elements = o.nonPriorityKeys
+	} else if o.EjectPriority {
+		elements = o.PriorityGroups[len(o.PriorityGroups)-1]
```

Comment:
If `o.PriorityGroups` is empty, we risk a panic here in line 34. Can we add a `len(o.PriorityGroups) > 0`  check?